tests/unit/base/trained_model_serialization_unittest.cc.jinja2

```diff
@@ -0,0 +1,210 @@
+				X_test_bin(1, i)=CMath::random(-1.0, 3.0);
+			}
+
+			feat_train_bin=new CDenseFeatures<float64_t>(X_bin);
+			feat_test_bin=new CDenseFeatures<float64_t>(X_test_bin);
+			label_train_bin=new CBinaryLabels(Y_bin);
+			SG_REF(feat_train_bin);
+			SG_REF(feat_test_bin);
+			SG_REF(label_train_bin);
+
+			/* regression data */
+			SGMatrix<float64_t> X_reg(2, N_TRAIN);
+			SGMatrix<float64_t> X_test_reg(2, N_TEST);
+			SGVector<float64_t> Y_reg(N_TRAIN);
+
+			for (index_t i=0; i<N_TRAIN; ++i)
+			{
+				X_reg(0,i)=CMath::random(-1.0, 1.0);
+				X_reg(1,i)=CMath::random(-1.0, 1.0);
+				Y_reg[i]=3*X_reg(0,i)+4*X_reg(1,i);
+			}
+
+			for (index_t i=0; i<N_TEST; ++i)
+			{
+				X_test_reg(0,i)=CMath::random(-1.0, 1.0);
+				X_test_reg(1,i)=CMath::random(-1.0, 1.0);
+			}
+
+			feat_train_reg=new CDenseFeatures<float64_t>(X_reg);
+			feat_test_reg=new CDenseFeatures<float64_t>(X_test_reg);
+			label_train_reg=new CRegressionLabels(Y_reg);
+			SG_REF(feat_train_reg);
+			SG_REF(feat_test_reg);
+			SG_REF(label_train_reg);
+		}
+
+		static void TearDownTestCase() {
+			SG_UNREF(feat_train_bin);
+			SG_UNREF(feat_test_bin);
+			SG_UNREF(label_train_bin);
+
+			SG_UNREF(feat_train_reg);
+			SG_UNREF(feat_test_reg);
+			SG_UNREF(label_train_reg);
+		}
+
+		static const index_t N_TRAIN=30;
+		static const index_t N_TEST=10;
+
+		static CDenseFeatures<float64_t>* feat_train_bin;
+		static CDenseFeatures<float64_t>* feat_test_bin;
+		static CBinaryLabels* label_train_bin;
+
+		static CDenseFeatures<float64_t>* feat_train_reg;
+		static CDenseFeatures<float64_t>* feat_test_reg;
+		static CRegressionLabels* label_train_reg;
+};
+
+CDenseFeatures<float64_t>* TrainedModelSerializationTest::feat_train_bin;
+CDenseFeatures<float64_t>* TrainedModelSerializationTest::feat_test_bin=NULL;
+CBinaryLabels* TrainedModelSerializationTest::label_train_bin=NULL;
+
+CDenseFeatures<float64_t>* TrainedModelSerializationTest::feat_train_reg=NULL;
+CDenseFeatures<float64_t>* TrainedModelSerializationTest::feat_test_reg=NULL;
+CRegressionLabels* TrainedModelSerializationTest::label_train_reg=NULL;
+
+{% macro test(class, problem_type) -%}
+TEST_F(TrainedModelSerializationTest, {{class}})
+{
+{% if problem_type == 'binary' %}
+	CDenseFeatures<float64_t>* feat_train=feat_train_bin;
+	CDenseFeatures<float64_t>* feat_test=feat_test_bin;
+	CBinaryLabels* label_train=label_train_bin;
+{% elif problem_type == 'regression' %}
+	CDenseFeatures<float64_t>* feat_train=feat_train_reg;
+	CDenseFeatures<float64_t>* feat_test=feat_test_reg;
+	CRegressionLabels* label_train=label_train_reg;
+{% endif %}
+
+	/* instantiate and train the machine */
+	auto machine=new C{{class}}();
+	machine->set_features(feat_train);
+	machine->set_labels(label_train);
+	bool train_success=machine->train();
+	ASSERT_TRUE(train_success);
+
+	/* predict with trained machine */
+	CLabels* predictions=machine->apply(feat_test);
+//	SGVector<float64_t> prediction_vector=predictions->get_labels();
```

Comment:
I think you can just use `equals`, so no need to actually access the labels. All done via polymorphism